cmd/mesh/upgrade.go

```diff
@@ -0,0 +1,353 @@
+// the cluster by given kubeConfigPath and context
+func applyUpgradeManifest(inFilename, kubeConfigPath, context string, dryRun, verbose bool) {
+	genApplyManifests(nil, inFilename, dryRun,
+		verbose, kubeConfigPath, context, upgradeWaitSecWhenApply, l)
+}
+
+// checkUpgradeValues checks the upgrade eligibility by comparing the current values with the target values
+func checkUpgradeValues(curValues string, tarValues string, yes bool) {
+	diff := compare.YAMLCmp(curValues, tarValues)
+	if diff == "" {
+		l.logAndPrintf("Upgrade check: Values unchanged. The target values are identical to the current values.\n")
+	} else {
+		l.logAndPrintf("Upgrade check: Warning!!! the following values will be changed as part of upgrade. "+
+			"If you have not overridden these values, they will change in your cluster. Please double check they are correct:\n%s", diff)
+	}
+
+	if yes {
+		return
+	}
+	if !confirm("Confirm to proceed [y/N]?", os.Stdout) {
+		l.logAndFatalf("Abort.")
+	}
+}
+
+// genICPSFromFile generates an IstioControlPlaneSpec for a spec file
+func genICPSFromFile(filename string, force bool) *v1alpha2.IstioControlPlaneSpec {
+	_, overlayICPS, err := genICPS(filename, "", "", force, l)
+	if err != nil {
+		l.logAndFatalf("Failed to generate ICPS from file %s, error: %s",
+			filename, err)
+	}
+	return overlayICPS
+}
+
+// genValuesFromFile generates values for a spec file
+func genValuesFromFile(filename string, force bool) string {
+	values, err := genProfile(true, filename, "", "", "", force, l)
+	if err != nil {
+		l.logAndFatalf("Abort. Failed to generate values from file: %v, error: %v", filename, err)
+	}
+	return values
+}
+
+// readValuesFromInjectorConfigMap reads the values from the config map of sidecar-injector.
+func readValuesFromInjectorConfigMap(kubeClient manifest.ExecClient, istioNamespace string) string {
+	configMapList, err := kubeClient.ConfigMapForSelector(istioNamespace, "istio=sidecar-injector")
+	if err != nil || len(configMapList.Items) == 0 {
+		l.logAndFatalf("Abort. Failed to retrieve sidecar-injector config map: %v", err)
+	}
+
+	jsonValues := ""
+	foundValues := false
+	for _, item := range configMapList.Items {
+		if item.Name == "istio-sidecar-injector" && item.Data != nil {
+			jsonValues, foundValues = item.Data["values"]
+			if foundValues {
+				break
+			}
+		}
+	}
+
+	if !foundValues {
+		l.logAndFatalf("Abort. Failed to find values in sidecar-injector config map: %v", configMapList)
+	}
+
+	yamlValues, err := yaml.JSONToYAML([]byte(jsonValues))
+	if err != nil {
+		l.logAndFatalf("jsonToYAML failed to parse values:\n%v\nError:\n%v", yamlValues, err)
+	}
+
+	return string(yamlValues)
+}
+
+// checkSupportedVersions checks if the upgrade cur -> tar is supported by the tool
+func checkSupportedVersions(cur, tar, versionsURI string) {
+	if cur == tar {
+		l.logAndFatalf("Abort. The current version %v equals to the target version %v.", cur, tar)
```

Comment:
Per comment above, should be returning error and logging higher in the stack so that controller code can use this.

Reply:
Sure, moved.